pkg/image/controller/trigger/trigger_controller.go

```diff
@@ -0,0 +1,399 @@
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		glog.V(4).Infof("Error syncing image stream %v: %v", key, err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	glog.V(4).Infof("Dropping image stream %q out of the queue: %v", key, err)
+	c.queue.Forget(key)
+}
+
+// resourceWorker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (c *TriggerController) resourceWorker() {
+	for c.processNextResource() {
+	}
+	glog.V(4).Infof("Resource worker stopped")
+}
+
+func (c *TriggerController) processNextResource() bool {
+	key, quit := c.imageChangeQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.imageChangeQueue.Done(key.(string))
+
+	err := c.syncResourceFn(key.(string))
+	c.handleResourceErr(err, key.(string))
+
+	return true
+}
+
+func (c *TriggerController) handleResourceErr(err error, key string) {
+	if err == nil {
+		c.imageChangeQueue.Forget(key)
+		return
+	}
+
+	if delay, ok := c.resourceFailureDelayFn(c.imageChangeQueue.NumRequeues(key)); ok {
+		glog.V(4).Infof("Error syncing resource %s: %v", key, err)
+		c.imageChangeQueue.AddAfter(key, delay)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	glog.V(4).Infof("Dropping resource %q out of the queue: %v", key, err)
+	c.imageChangeQueue.Forget(key)
+}
+
+// syncImageStream will sync the image stream with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (c *TriggerController) syncImageStream(key string) error {
+	if glog.V(4) {
+		startTime := time.Now()
+		glog.Infof("Started syncing image stream %q", key)
+		defer func() {
+			glog.Infof("Finished syncing image stream %q (%v)", key, time.Now().Sub(startTime))
+		}()
+	}
+
+	// find the set of triggers to act on
+	triggered, err := c.triggerCache.ByIndex("images", key)
+	if err != nil {
+		return err
+	}
+	if len(triggered) == 0 {
+		return nil
+	}
+
+	// queue every trigger
+	// TODO: possibly filter for impossible triggers
+	for _, t := range triggered {
+		entry := t.(*trigger.CacheEntry)
+		c.imageChangeQueue.Add(entry.Key)
+	}
+	return nil
+}
+
+// syncResource handles a set of changes against one of the possible resources generically.
+func (c *TriggerController) syncResource(key string) error {
+	if glog.V(1) {
+		//startTime := time.Now()
+		glog.Infof("Started syncing resource %q", key)
+		defer func() {
+			glog.Infof("Finished syncing resource %q", key) //, time.Now().Sub(startTime))
+		}()
+	}
+
+	parts := strings.SplitN(key, "/", 2)
+	source := c.triggerSources[parts[0]]
+	obj, exists, err := source.Store.GetByKey(parts[1])
+	if err != nil {
+		return fmt.Errorf("unable to retrieve %s %s from store: %v", parts[0], parts[1], err)
+	}
+	if !exists {
+		return nil
+	}
+
+	if err := source.Reaction.ImageChanged(obj, c.tagRetriever); err != nil {
+		return err
```

Comment:
We could maybe fire an event if the trigger is invalid.  I don't like that in general because resync would fire a new event.  Another option is to mutate the annotation.